linkerd/http-route/src/http/filter/inject_failure.rs

```diff
@@ -0,0 +1,74 @@
+use std::hash::Hash;
+
+/// A filter that responds with an HTTP error.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct InjectFailure {
+    pub response: FailureResponse,
+    pub distribution: Distribution,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct FailureResponse {
+    pub status: http::StatusCode,
+    pub message: std::sync::Arc<str>,
+}
+
+/// A Bernoulli distribution that implements `Hash`, `PartialEq`, and `Eq`.
+#[derive(Clone, Debug)]
+pub struct Distribution {
+    numerator: u32,
+    denominator: u32,
+    inner: rand::distributions::Bernoulli,
+}
+
+// === impl InjectFailure ===
+
+impl InjectFailure {
+    pub fn apply(&self) -> Option<FailureResponse> {
+        use rand::distributions::Distribution;
+
+        if self.distribution.sample(&mut rand::thread_rng()) {
+            return Some(self.response.clone());
+        }
+
+        None
+    }
+}
+
+// === impl InjectFailure ===
+
+impl Distribution {
+    pub fn from_ratio(
+        numerator: u32,
+        denominator: u32,
+    ) -> Result<Self, rand::distributions::BernoulliError> {
+        let inner = rand::distributions::Bernoulli::from_ratio(numerator, denominator)?;
+        Ok(Self {
+            numerator,
+            denominator,
+            inner,
+        })
+    }
+}
+
+impl rand::distributions::Distribution<bool> for Distribution {
+    #[inline]
+    fn sample<R: rand::Rng + ?Sized>(&self, rng: &mut R) -> bool {
+        self.inner.sample(rng)
+    }
+}
+
+impl PartialEq for Distribution {
+    fn eq(&self, other: &Self) -> bool {
+        self.numerator == other.numerator && self.denominator == other.denominator
+    }
+}
+
+impl Eq for Distribution {}
+
+impl Hash for Distribution {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.numerator.hash(state);
+        self.denominator.hash(state);
+    }
+}
```

Comment:
not *totally* sure why we need our own `Distribution` type when it appears to just be a wrapper over `rand`'s `distributions::Bernoulli` type --- is this just so we can impl `Hash` on it, or something? maybe worth a comment indicating that...

Reply:
https://github.com/linkerd/linkerd2-proxy/blob/e8220da644fb4fc5e01b386c256cc7905a0eacd5/linkerd/http-route/src/http/filter/inject_failure.rs#L16